7093691: Nimbus LAF: disabled JComboBox using renderer has bad font color

DamonGuy opened the pull request:
Before the fix, a JComboBox in Nimbus L&F would have normal black text even when the JComboBox was disabled if SynthComboBoxRenderer was replaced with a DefaultListCellRenderer. This text should be greyed out like in other L&F's. When looking into the defaults for Nimbus L&F files for attributes and states of a JComboBox, it confirm that the intention for disabled JComboBoxes is nimbusDisabledText (which is grey text).

SynthComboBoxes have an additional check in its default SynthComboBoxRenderer that enables/disables the renderer itself. The SynthComboBoxRenderer inherits its enabled state from the parent ComboBox. Since the renderer with DefaultListCellRenderer is in a separate class without a reference to the comboBox, a listener was added to SynthComboBoxUI.

An additional issue occurred in DefaultListCellRenderer because the renderer overrode the listener's re-assigned enabled state. In testing, setting the enabled state in DefaultListCellRenderer is redundant for all L&F's and is not needed here. However, instead of removing it altogether, a conditional was added specifically to allow ComboBoxes to skip setting enabled state here.

After the fix, the Nimbus JComboBox with DLCR set matches the appearance of a normal Nimbus JComboBox. I can enable/disable the JComboBoxes in the test, and the UI elements behave and appear as expected.

DamonGuy commented:
Here are screenshots of the bug and the UI elements after the fix is applied.

Disabled JComboBox with DLCR before fix:
<img width="644" alt="Screen Shot 2023-02-01 at 3 52 46 PM" src="https://user-images.githubusercontent.com/96267980/216385047-a9bde4f9-c33d-430b-b68f-f4a9df8bf8a9.png">

After fix:
<img width="643" alt="Screen Shot 2023-02-01 at 3 46 19 PM" src="https://user-images.githubusercontent.com/96267980/216384982-1407f271-d28d-423f-843b-5e1c59ba984c.png">

kumarabhi006 commented on test/jdk/javax/swing/JComboBox/DisabledComboBoxFontTest.java:
```diff
@@ -0,0 +1,138 @@
+
+        hasNimbus = false;
+
+        for (LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
```
It seems the intention to run the test is in Nimbus LAF only. I think you can directly set the Nimbus LAF using `UIManager.setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");` and can avoid the `for-loop, condition check` and `hasNimbus `variable.

To print the error message, LAF can be set within try-catch block.

DamonGuy replied:
Thanks for the suggestion. I made the change. I initially had this loop to test locally for all LAF's, but ended up using this test for Nimbus only as originally designed.

kumarabhi006 commented on test/jdk/javax/swing/JComboBox/DisabledComboBoxFontTest.java:
```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
```
Copyright year can be changed to 2023.

DamonGuy replied:
Updated

DamonGuy commented on test/jdk/javax/swing/JComboBox/DisabledComboBoxFontTest.java:
```diff
@@ -0,0 +1,130 @@
+            "status with the button on the right of the frame.";
+
+    private static JFrame frame;
+    private static boolean hasNimbus;
```
Thanks, updated

DamonGuy commented:
@prsadhuk could you review this as well since I got your input about this issue previously? 

Initially, I thought it was the default value of the disabled combobox, but none of the skin.laf UI states/values affected the displayed text color.

So, it may be a state issue as you suggested. I checked the enabled state of each L&F, and the combobox's state is correct, but I noticed that the renderer in Synth has an additional condition that sets the renderer to disabled/enabled to match the combobox state. In testing, setting the renderer to disabled produced the correct behavior, so I made a listener as a workaround to access the combobox's renderer for the case with a DefaultListCellRenderer.

With the test I created, this approach creates the correct behavior, so I implemented the fix the cleanest way possible.

honkar-jdk reviewed (approved):
Tested the fix on SwingSet2 and the fix works as expected. 
Copyright year needs to be updated for SynthComboBoxUI , DefaultListCellRenderer too.

prsadhuk commented:
https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthComboBoxUI.java#L552-L558
The present code SynthComboBoxRenderer  does enable/disable the listcell renderer based on comboBox status, why it is not working? why do we need to use a listener for DefaultListCellRenderer?

Also, I think it will be better to make the test automated..something like draw comboBox content into BufferedImage and pixel compare if it is same for disabled/enabled and fail if they are same

DamonGuy commented:
> https://github.com/openjdk/jdk/blob/98433a2f6e7fe97e03ed26673c9925d7b26466bf/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthComboBoxUI.java#L552-L558
> 
> The present code SynthComboBoxRenderer does enable/disable the listcell renderer based on comboBox status, why it is not working? why do we need to use a listener for DefaultListCellRenderer?
> 
> Also, I think it will be better to make the test automated..something like draw comboBox content into BufferedImage and pixel compare if it is same for disabled/enabled and fail if they are same and iterate for all L&Fs as it will also affect GTK as per the comment

The setEnabled line you mentioned in SynthComboBoxUI applies to SynthComboBoxRenderer only. This issue only occurs when DefaultListCellRenderer is used for a disabled Nimbus ComboBox instead of the default renderer for Nimbus (SynthComboBoxRenderer). So this code is never reached.

I will work on making the test automated instead if preferred.

DamonGuy commented:
> Also, I think it will be better to make the test automated..something like draw comboBox content into BufferedImage and pixel compare if it is same for disabled/enabled and fail if they are same and iterate for all L&Fs as it will also affect GTK as per the comment

I created an automated test, but for some L&F's (such as Nimbus), an enabled vs disabled ComboBox differs in more than just the text color. The background, border, and button color may change as well. So, even if I shave off the dimensions of the border and button, it seems difficult to detect just the text for comparison.

Before the fix, the issue was that the ComboBox as a whole worked correctly for disabled ComboBoxes with a DLCR, but the text was black instead of grey. This was the only difference.

A manual test might still be the best approach since more aspects than just the test color can be checked. So, I think it's best to not automate this test.

prsadhuk commented:
`SynthComboBoxUI.createRenderer` points to `SynthComboBoxRenderer `so I think `DefaultListCellRenderer `should not come into picture for rendering cells in the list which should go via `getListCellRendererComponent`.
If it is playing a part, then how it works in other L&F which also has it own renderer (for example, `BasicComboBoxUI.createRenderer` points to `BasicComboBoxRenderer`, AquaComboBox points to `AquaComboBoxRenderer`) and they dont have this listener to make it work, so the question why do we need a listener for Synth still not clear to me..

prsadhuk commented:
> I created an automated test, but for some L&F's (such as Nimbus), an enabled vs disabled ComboBox differs in more than just the text color. The background, border, and button color may change as well

Then can we not compare the comboBox without DLCR with comboBox with DLCR when both are disabled...both should be same as per your screenshot

alisenchung commented:
I ran the test on linux on GTK L&F. Screenshots of enabled/disabled JComboBoxes: 
![Screenshot 2023-02-08 125006](https://user-images.githubusercontent.com/90066231/217648359-1caf92ae-5df7-4812-b2a8-dc8669d49235.png)
![Screenshot 2023-02-08 125030](https://user-images.githubusercontent.com/90066231/217648362-4bae9ddb-8a5a-4f01-a053-95be832d458b.png)

They look fine to me.

DamonGuy commented:
> > I created an automated test, but for some L&F's (such as Nimbus), an enabled vs disabled ComboBox differs in more than just the text color. The background, border, and button color may change as well
> 
> Then can we not compare the comboBox without DLCR with comboBox with DLCR when both are disabled...both should be same as per your screenshot

A comboBox with DLCR is slightly smaller than one without DLCR (138x25 vs 140x25). If I force the same dimensions with setSize, the RGB values are slightly off. If these comboBoxes are supposed to exactly match in dimensions, they currently don't, and this is due to what would be a separate bug if so.

DamonGuy commented:
> `SynthComboBoxUI.createRenderer` points to `SynthComboBoxRenderer `so I think `DefaultListCellRenderer `should not come into picture for rendering cells in the list which should go via `getListCellRendererComponent`. If it is playing a part, then how it works in other L&F which also has it own renderer (for example, `BasicComboBoxUI.createRenderer` points to `BasicComboBoxRenderer`, AquaComboBox points to `AquaComboBoxRenderer`) and they dont have this listener to make it work, so the question why do we need a listener for Synth still not clear to me..

I thought the same, but Nimbus seems to act different for this instance only. I traced the conditional in SynthComboBoxRenderer, and this was added in Apr 25, 2009 [here](https://github.com/openjdk/jdk/commit/c738ebdc6789068c3faf9e656bd4c02f84fe82bc#diff-54280d4d928e7579aa04324bb7d0a306f621b92d77d0dc508d8a269c124664a8). This seems like the introduction of Nimbus and how it's enabled/disabled ComboBoxes for Nimbus, with the oversight of setting a different renderer. This looks like the only OS that does this like you mentioned. The reason why this need to setEnabled in the renderer for Nimbus wasn't stated in the comment or bug description page. But it does seem to still need it as this was the only fix I found.

prsadhuk commented:
> > > I created an automated test, but for some L&F's (such as Nimbus), an enabled vs disabled ComboBox differs in more than just the text color. The background, border, and button color may change as well
> > 
> > 
> > Then can we not compare the comboBox without DLCR with comboBox with DLCR when both are disabled...both should be same as per your screenshot
> 
> A comboBox with DLCR is slightly smaller than one without DLCR (138x25 vs 140x25). If I force the same dimensions with setSize, the RGB values are slightly off. If these comboBoxes are supposed to exactly match in dimensions, they currently don't, and this is due to what would be a separate bug if so.

maybe you can try to compare the mid scanline rather than the whole rectangle

DamonGuy commented:
> > > > I created an automated test, but for some L&F's (such as Nimbus), an enabled vs disabled ComboBox differs in more than just the text color. The background, border, and button color may change as well
> > > 
> > > 
> > > Then can we not compare the comboBox without DLCR with comboBox with DLCR when both are disabled...both should be same as per your screenshot
> > 
> > 
> > A comboBox with DLCR is slightly smaller than one without DLCR (138x25 vs 140x25). If I force the same dimensions with setSize, the RGB values are slightly off. If these comboBoxes are supposed to exactly match in dimensions, they currently don't, and this is due to what would be a separate bug if so.
> 
> maybe you can try to compare the mid scanline rather than the whole rectangle

Replaced the original test with an automated version using a mid scanline. I can't scan across the whole midline because of the width difference, so I adjusted the testing method slightly to accommodate for this.

prsadhuk commented:
Did you try extending DLCR for `SynthComboBoxRenderer `instead of implementing LCR interface?
https://github.com/openjdk/jdk/blob/1480d418e3b7d1f36ace24a043a273fca446eefa/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthComboBoxUI.java#L502

just as it is done here
https://github.com/openjdk/jdk/blob/1480d418e3b7d1f36ace24a043a273fca446eefa/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthListUI.java#L211

DamonGuy commented:
> extends DefaultListCellRenderer.UIResource

I tried implementing this after your suggestion, but it doesn't achieve the wanted result. This makes sense from my understanding because with a new comboBox set to using a new DefaultListCellRenderer, the SynthComboBoxRenderer wouldn't be used. Changing what is extended for SynthComboBoxRenderer would only affect the normal comboBox using the default Synth renderer. The listener is the way around this because the comboBox with a DLCR still is a SynthComboBox but does not rely on the SynthComboBoxRenderer.

azuev-java commented on test/jdk/javax/swing/JComboBox/DisabledComboBoxFontTestAuto.java:
```diff
@@ -0,0 +1,176 @@
+                + "/" + lafName + "EnabledDLCR.png"));
+        ImageIO.write(disabledImage2, "png", new File(testDir
+                + "/" + lafName + "DisabledDLCR.png"));
+        System.out.println("DIR: " + testDir);
```
Is this needed? I know it is just a test but please remove the debug output when not needed.

DamonGuy replied:
Removed the message. I added it to double check the saved images' file path, but didn't remove it before uploading.

azuev-java commented on test/jdk/javax/swing/JComboBox/DisabledComboBoxFontTestAuto.java:
```diff
@@ -0,0 +1,176 @@
+            SwingUtilities.invokeAndWait(() -> setLookAndFeel(laf));
+            SwingUtilities.invokeAndWait(DisabledComboBoxFontTestAuto::createCombo);
+            SwingUtilities.invokeAndWait(DisabledComboBoxFontTestAuto::paintCombo);
+            testMethod();
```
So test will fail whenever it encounters the first wrong color? Ideally we would run test for all the installed LAFs anyways and report which ones are failed. But in this case you would need to pass the LAF name to the test method so it saves screenshots for all the LAFs for future analysis.

DamonGuy replied:
I added a StringBuffer which now outputs the failed LAFs to the thrown runtime exception.

DamonGuy commented:
@prsadhuk @honkar-jdk Could you take another look because of the updates?

prsadhuk commented:
It seems the test fails on windows for WIndows L&F..This is what I got
```
./jdk/bin/java DisabledComboBoxFontTestAuto
PASSED
PASSED
PASSED
Windows Enabled RGB failure: 240, 240, 240 vs 225, 225, 225
PASSED
Exception in thread "main" java.lang.RuntimeException: FAIL - Enabled and disabled ComboBox does not match in these LAFs: Windows
        at DisabledComboBoxFontTestAuto.main(DisabledComboBoxFontTestAuto.java:178)
```

It seems it does not fail in mach5 maybe because it is not run in hidpi system.

Also, I was expecting some explanation about this snippet if it is still needed for GTK (because of which, I think, you had to do the change in `javax/swing/DefaultListCellRenderer.java`)

https://github.com/openjdk/jdk/blob/21a6ab1e3ea5228a31955d58fe75e5ae66d1c6cd/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthComboBoxUI.java#L552-L560

You took care of `enabled `property..If the above snippet still hold significance, do we need to do the same for `orientation `property too

DamonGuy commented:
> https://github.com/openjdk/jdk/blob/21a6ab1e3ea5228a31955d58fe75e5ae66d1c6cd/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthComboBoxUI.java#L552-L560
> 
> You took care of `enabled `property..If the above snippet still hold significance, do we need to do the same for `orientation `property too

I tested without any additional listener for orientation and set the combo's orientation to right_to_left. It's working correctly. So, I tested the code block you referenced above and removed the code for updating orientation when `comboBox != null`. This still works correctly. So, this portion of the code that updates the orientation in SynthComboBoxRenderer may be redundant.

However, the enabled portion seems to be needed, and the additional listener added is needed to generate the correct text color.

Nimbus test with a normal comboBox and a custom comboBox (with DefaultListCellRenderer) set with orientation RTL:
<img width="435" alt="Screen Shot 2023-03-22 at 9 44 25 AM" src="https://user-images.githubusercontent.com/96267980/227014415-3920c785-e2ea-45ba-b828-39e20a0d1d4d.png">


Test with a disabled, normal comboBox and custom comboBox (with DefaultListCellRenderer) set with orientation RTL:
<img width="432" alt="Screen Shot 2023-03-22 at 10 00 23 AM" src="https://user-images.githubusercontent.com/96267980/227014447-3a3f0aa0-377c-4f6b-9448-100d167652f7.png">

prsadhuk reviewed (approved):
ok..

DamonGuy commented:
/integrate

mrserb commented:
> It seems the test fails on windows for WIndows L&F..This is what I got
> It seems it does not fail in mach5 maybe because it is not run in hidpi system.

@prsadhuk Did you able to fix it, or report the new JBS bug for that? I have the same issue, the test fails on the HiDPI system.
